src/diagram/models/variable.ts

```diff
@@ -43,117 +43,265 @@ export const Variable = types.model("Variable", {
+      // errors MathJS provides a data property on the error that includes
+      // the character location and more info about the error.
+      if (e.message?.startsWith("Units do not match")) {
+        return {error: "incompatible units"};
+      } else if (e.message?.startsWith("Unexpected type of argument")) {
+        // This can happen when a unit-less value is added or subtracted from a
+        // value with a unit. We could provide more information about this if we
+        // want to. When supporting generic expressions we probably will want to.
+        return {error: "incompatible units"};
       } else {
-        return {error: "no operation"};
+        return {error: `unknown error: ${e.message}`};
       }
     }
-    return {value: self.value};
   },
   
   // If there are two inputs then units can't be changed
   // otherwise current node units override previous node units
   get computedUnitIncludingMessageAndError(): {unit?: string, error?: string, message?: string} {
-    if (self.inputA && self.inputB) {
-      if (self.operation) {
-        // If there is no unit, then use "1", that way the simplification of multiplication
-        // and division will work properly
-        // @ts-expect-error THIS
-        const inputAUnit = this.inputA.computedUnit;
+    if (self.numberOfInputs === 0) {
+      // Just return the current unit. 
+      // The current unit might be undefined
+      return {unit: self.unit};
+    }
+
+    const expression = self.expression;
+    if (!expression) {
+      // We should only have a empty expression if there are two inputs and the
+      // operation is not set. 
+      // With a single input the expression should always be set.
+      // The computedValue above will already include a message above so it is
+      // not reported here too.
+      return {};
+    }
+
+    // @ts-expect-error THIS
+    const inputAMathValue = this.inputA?.mathValueWithValueOr1;
+
+    // The input value could be a number so it could be 0 which would be
+    // valid
+    if (self.inputA && inputAMathValue === undefined) {
+      // Because we are using `1` if there is no input value, and we just
+      // return the value if there is no unit, the only reason the
+      // inputAMathValue should be undefined is if there is an error with the
+      // input unit.
+      return {error: "invalid input units"};
+    }
+
+    // @ts-expect-error THIS
+    const inputBMathValue = this.inputB?.mathValueWithValueOr1;
+
+    // The input value could be a number so it could be 0 which would be
+    // valid
+    if (self.inputB && inputBMathValue === undefined) {
+      // Because we are using `1` if there is no input value, and we just
+      // return the value if there is no unit, the only reason the
+      // inputAMathValue should be undefined is if there is an error with the
+      // input unit.
+      return {error: "invalid input units"};
+    }
```

Comment:
There's less duplication here, but as before this could be DRYed up with a `validateInputValueOr1` function. This would pay dividends down the road when there could be arbitrary numbers of inputs.

Reply:
This should be better now.